### PR TITLE
Add the ability to delete nodes + paths, user logging, and steps to update

### PR DESCRIPTION
**Pull request by mawburn**
I figured actually changing the updates was too much for this, but this PR adds the ability to delete nodes and adds logging for users. Which is just thrown into a table for now but that should be fine. There aren't any indexes or keys on the logging table.

### CODE REVIEW

**Comment by mawburn** [comment]
I also started the ability to separate updates, but I think that should be another PR. I can use the endpoint I have here as a base to build from

File: packages/api-server/db_migrations/004_logging.sql [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+CREATE TABLE IF NOT EXISTS user_logs (
+  user_id INT NOT NULL,
+  server_id INT NOT NULL,
+  user_action user_action_enum NOT NULL,
+  details VARCHAR (1024),
+  created_on TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT NOW()
+);
+
+ALTER TABLE portals ADD COLUMN updated_by INT;
+ALTER TABLE portals ADD COLUMN updated_ON TIMESTAMP WITH TIME ZONE;
```
Comment (Tebro) [inline packages/api-server/db_migrations/004_logging.sql:12]: Can't these two be fetched from the new user_logs table?

File: packages/api-server/src/api/portal.ts [hunk 2]
```diff
@@ -102,4 +149,48 @@ router.post('/', async (req, res) => {
...
+
+    await deleteServerPortal(portalIds, req.userId, req.serverId)
+    res.send(204)
+  } catch (err) {
+    logger.log.error('Unable to delete', err)
+    res.status(500).send({ error: 'Error deleting portal' })
+  }
+})
+
+// TODO update in later PR
```
Comment (Tebro) [inline packages/api-server/src/api/portal.ts:174]: What?

File: packages/api-server/src/api/portal.ts [hunk 3]
```diff
@@ -102,4 +149,48 @@ router.post('/', async (req, res) => {
...
+router.put('/:id(\\d+)', async (req, res) => {
+  try {
+    const body = req.body
+    const expireTime = getExpireTime(body.size, body.hours, body.minutes)
+    const conns = body.connection.sort()
+
+    await updateServerPortal(
+      Number(req.params.id),
```
Comment (Tebro) [inline packages/api-server/src/api/portal.ts:182]: what is wrong with parseInt?

File: packages/discord-bot/package.json [hunk 4]
```diff
@@ -37,6 +37,6 @@
     "@types/node-fetch": "^2.5.7",
     "@types/pg": "^7.14.6",
     "ts-node-dev": "^1.0.0",
-    "typescript": "^4.0.5"
+    "typescript": "~3.7.2"
```
Comment (Tebro) [inline packages/discord-bot/package.json:40]: Why the downgrade?

File: packages/frontend/src/PortalMap/ControlBar/DeleteNode.tsx [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+      if (deleteZone) {
+        deletePortals(edgeData)
+      }
+
+      setIsOpen(false)
+    },
+    [deletePortals, edgeData]
+  )
+
+  return edgeData.length > 0 ? (
```
Comment (Tebro) [inline packages/frontend/src/PortalMap/ControlBar/DeleteNode.tsx:36]: Use `edgeData.length > 0 &&` ?

File: packages/frontend/src/PortalMap/index.tsx [hunk 6]
```diff
@@ -72,13 +80,24 @@ const PortalMap = () => {
     [zones, activeZoneName]
   )
 
+  const [activeZoneEdgeData, setActiveZoneEdgeData] = useState<CytoEdgeData[]>(
+    []
```
Comment (Tebro) [inline packages/frontend/src/PortalMap/index.tsx:84]: If this is some prettify rule then I think they need tweaking

File: shared/data-models/package.json [hunk 7]
```diff
@@ -23,7 +23,7 @@
     "@types/node": "^14.14.7",
     "@types/pg": "^7.14.6",
     "@types/redis": "^2.8.28",
-    "typescript": "^4.0.5"
+    "typescript": "~3.7.2"
```
Comment (Tebro) [inline shared/data-models/package.json:26]: Another downgrade?

File: shared/data-models/src/models/User.ts [hunk 8]
```diff
@@ -267,6 +273,20 @@ export default class UserModel {
     }
   }
 
+  logUserAction = async (
+    userId: number,
+    serverId: number,
+    action: UserAction,
+    details: string
+  ) =>
+    await this.query(
```
Comment (Tebro) [inline shared/data-models/src/models/User.ts:282]: No need to async/await this, just return the promise from query

File: packages/discord-bot/package.json (hunk 4 above)
Comment (mawburn) [reply packages/discord-bot/package.json:40]: It caused issues with VSC. React is locked at this version. So I made all the versions the same.

CRA 4 is out and I attempted to upgrade to that once but it was a lot of work and broke a lot of stuff so I had to roll back. I actually forgot to check in before upgrading and lost what I was working on that day.

File: packages/frontend/src/PortalMap/ControlBar/DeleteNode.tsx (hunk 5 above)
Comment (mawburn) [reply packages/frontend/src/PortalMap/ControlBar/DeleteNode.tsx:36]: Oops yeah. I had just a falsy check and was printing a 0. I was getting tired and it looks like I fixed the problem 2 different ways.

File: shared/data-models/src/models/User.ts (hunk 8 above)
Comment (mawburn) [reply shared/data-models/src/models/User.ts:282]: True

Comment (mawburn) [reply shared/data-models/src/models/User.ts:282]: True

File: packages/api-server/src/api/portal.ts (hunk 3 above)
Comment (mawburn) [reply packages/api-server/src/api/portal.ts:182]: Nothing. Number is a little cleaner to write and I like it better.

File: packages/api-server/src/api/portal.ts (hunk 2 above)
Comment (mawburn) [reply packages/api-server/src/api/portal.ts:174]: I explained at the beginning of the PR.

File: packages/api-server/db_migrations/004_logging.sql (hunk 1 above)
Comment (mawburn) [reply packages/api-server/db_migrations/004_logging.sql:12]: Yeah. I think I just added these out of habit. I don't even remember updating them.

File: packages/api-server/src/api/portal.ts (hunk 3 above)
Comment (mawburn) [reply packages/api-server/src/api/portal.ts:182]: https://thisthat.dev/number-constructor-vs-parse-int/

This article is what turned me on to it.

Comment (Tebro) [reply packages/api-server/src/api/portal.ts:182]: After reading that I am more convinced on parseInt, the NaN cases are just bad for Number()

File: packages/api-server/src/api/portal.ts (hunk 2 above)
Comment (Tebro) [reply packages/api-server/src/api/portal.ts:174]: I see, remove this from these commits then and bring it in with the rest.

File: packages/api-server/src/api/portal.ts (hunk 3 above)
Comment (mawburn) [reply packages/api-server/src/api/portal.ts:182]: You can check for isNaN

Comment (Tebro) [reply packages/api-server/src/api/portal.ts:182]: It returns 0.

File: packages/frontend/src/PortalMap/ControlBar/DeleteNode.tsx (hunk 5 above)
Comment (mawburn) [reply packages/frontend/src/PortalMap/ControlBar/DeleteNode.tsx:36]: Screen Shot 2020-11-29 at 5 14 38 PM
Actually this is why it's not that.

**Pull request merged**